### ISSUE TITLE
Error with modules, format, pipe operator - function with deduced return type cannot be used

### ISSUE THREAD

**Issue by rnikander** [llvm/llvm-project#61912]
This seems to be a bug.

https://godbolt.org/z/o1o97ar4j

Something bad happening with C++20 modules + `<format>` causes this confusing error. Remove the `#include <format>` and it goes away.

```console
In module 'MathMod' imported from /app/main.cc:3:
PolyTri.cc:15:37: error: function 'operator|<std::vector<int> &, std::__range_adaptor_closure_t<std::__bind_back_t<std::ranges::views::__drop::__fn, std::tuple<int>>>>' with deduced return type cannot be used before it is defined
        auto wtf = this->dcel.faces | std::views::drop(1);           
```

**Comment by ChuanqiXu9** [comment]
Noted that the problem remains after I replace `format` to `ranges`: https://godbolt.org/z/1zs9346Tn. So it may doesn't matter with format but something in experimental library.

**Comment by dvirtz** [comment]
Still seeing this error in Clang 18.1.0.
Any workarounds?

**Comment by ChuanqiXu9** [comment]
It looks like the original reproducer doesn't work in upstream: https://godbolt.org/z/5cP4W8YrM

Would you like to provide a new reproducer? (And it will be better if it is reduced.)

**Comment by dvirtz** [comment]
It reproduces across a library boundary: https://godbolt.org/z/7zac84afd

**Comment by ChuanqiXu9** [comment]
Yeah, it looks like there indeed some problems in the compiler side. To push fixing the underlying issue, may we can try to get a reproducer with containing any system headers.

A workaround for users may be: https://godbolt.org/z/5srq6s1cd

Another suggestion is, given you're using libc++ in the reproducer, try to use the std module: https://libcxx.llvm.org/Modules.html

If you're using libstdc++ actually, I still suggesting to mock a std module by yourself. Due to https://clang.llvm.org/docs/StandardCPlusPlusModules.html#reduce-duplications

**Comment by rnikander** [comment]
> If you're using libstdc++ actually, I still suggesting to mock a std module by yourself. Due to

@ChuanqiXu9 I followed your link there for the advice on "reducing duplications". Do you know any tricks to deal with `#define` stuff? For example, I tried it with Vulkan, and it seems to work until I hit `VK_NULL_HANDLE` which is `#define`. 

```
module;
#include <vulkan/vulkan.h>
export module vulkan;

export using ::VkWin32SurfaceCreateInfoKHR;
export using ::vkDestroySurfaceKHR;
...
export using ::VK_NULL_HANDLE;   // <--- nope...
```

**Comment by ChuanqiXu9** [comment]
I didn't know it. It looks like some configuration is incorrect in vulkan or I misunderstand your problem.

**Comment by rnikander** [comment]
> I didn't know it. It looks like some configuration is incorrect in vulkan or I misunderstand your problem.

Vulkan is configured fine. I mean that I don't see how to re-export a symbol that is defined in the header file as a macro (`#define VK_NULL_HANDLE ...`). But maybe I shouldn't take this github issue off-topic. I can ask somewhere else. :)

**Comment by ChuanqiXu9** [comment]
but modules can't export macros by the design...
